src/Val.cc

```diff
@@ -3240,20 +3231,24 @@ StringVal* ValManager::GetEmptyString() const
 	return empty_string;
 	}
 
-PortVal* ValManager::GetPort(uint32_t port_num, TransportProto port_type) const
+PortVal* ValManager::GetPort(uint32_t port_num, TransportProto port_type)
 	{
 	if ( port_num >= 65536 )
 		{
 		reporter->Warning("bad port number %d", port_num);
 		port_num = 0;
 		}
 
-	auto rval = ports[port_type][port_num];
+	auto& slot = ports[port_type][port_num];
+	auto rval = slot;
+	if ( !rval )
+		slot = rval = new PortVal(PortVal::Mask(port_num, port_type));
```

Comment:
I have slight reservations about the performance cost of this change. The reason why all of the vals and such are created during startup is to avoid the performance hit of making a new one every time we find a new port number. It does cost a bit of time at startup, but most use cases won't be starting and stopping zeek a lot.

Reply:
I disagree, because creating 10 MB of very small objects surpasses my (subjective) threshold where it's worth it; but I'll force-push without this commit.

Reply:
After your review, I was a bit confused to find this piece of code:
https://github.com/zeek/zeek/blob/a394d2e0d709ff507859484ea862e9468cc8a5c8/src/Type.cc#L1508-L1525
This is similar to what I proposed for `ValManager::GetPort()`, but this function gets called for each and every `Val` object, i.e. *very* often (it gets called 266,754 times from inside `ValManager::ValManager()` alone). But it manages only 26 objects.

Reply:
> It gets called 266,754 times from inside ValManager::ValManager() alone

Sure, it's 2 bools, 4096 counts, 512 ints, and 65536 * 4 ports that get created in `ValManager()`.

> this function gets called for each and every Val object

It does, but a very large majority of those val objects a pre-allocated at startup so they should be having `Ref()` called on them instead of a completely new instance created if they're reused. For example, the `0` and `1` value `count` and `int` objects are used a ton, but only one should ever be created.